File: rubix/extensions/com.nubeio.rubixos/ui-src/provision/bc-api.ts
// `bc-api.ts` — thin, named wrappers over the bc_* tools + read templates.
// Each function just forwards to callTool / fetchTemplate from the shared
// extension api so callers never hand-spell tool ids or template names.

import { callTool, fetchTemplate } from "../api";
import { EXTENSION_ID } from "../types";
import { bumpRefresh } from "./refresh";
import type {
  AlarmRow,
  AssignPageInput,
  AssignPageResult,
  DeviceRow,
  LabelRender,
  LocationRow,
  LogRow,
  MutationResult,
  PageDeleteResult,
  PageRow,
  PointRow,
  ProvisionInput,
  ProvisionResult,
  ScannedIdentity,
  SiteRow,
  TemplateRow,
  TemplateYaml,
  WidgetRow,
} from "./bc-types";

const tool = (name: string) => `${EXTENSION_ID}.${name}`;

// Every write goes through here so the shared data version bumps as
// soon as the server confirms — sibling list tabs then re-fetch and
// converge without a page reload. `decode` is read-only (no bump).
function mutate<T>(toolId: string, params: unknown): Promise<T> {
  return callTool<T>(toolId, params).then((res) => {
    bumpRefresh();
    return res;
  });
}

/* ------------------------------- mutations ------------------------------- */

export function decode(barcode: string): Promise<ScannedIdentity> {
  return callTool<ScannedIdentity>(tool("bc_decode"), { barcode });
}

export function provision(input: ProvisionInput): Promise<ProvisionResult> {
  return mutate<ProvisionResult>(tool("bc_provision"), input);
}

// Place an already-commissioned (pending) device on a page: generates
// its widgets and flips status to provisioned. One of page_id / new_page
// is required. Mutation → bumps refresh so list tabs converge.
export function assignPage(input: AssignPageInput): Promise<AssignPageResult> {
  return mutate<AssignPageResult>(tool("bc_device_assign_page"), input);
}

export function deviceUpdate(
  row: { device_id: string } & Record<string, unknown>,
): Promise<MutationResult> {
  return mutate<MutationResult>(tool("bc_device_update"), { row });
}

export function decommission(
  device_ids: ReadonlyArray<string>,
  hard = false,
): Promise<MutationResult> {
  return mutate<MutationResult>(tool("bc_device_decommission"), { device_ids, hard });
}

export function siteCreate(row: { site_id: string; name: string }): Promise<MutationResult> {
  return mutate<MutationResult>(tool("bc_site_create"), { row });
}

export function locationCreate(
  row: { location_id: string; site_id: string; name: string },
): Promise<MutationResult> {
  return mutate<MutationResult>(tool("bc_location_create"), { row });
}

export function pageCreate(
  row: { page_id: string; name: string; site_id?: string; location_id?: string },
): Promise<MutationResult> {
  return mutate<MutationResult>(tool("bc_page_create"), { row });
}

// Update a page by `page_id` — rename, or re-pin its site/location.
export function pageUpdate(
  row: { page_id: string; name?: string; site_id?: string; location_id?: string },
): Promise<MutationResult> {
  return mutate<MutationResult>(tool("bc_page_update"), { row });
}

// Delete a page by `page_id`. Its widgets are dropped; devices placed on
// it are kept but detached (page_id cleared, status → pending).
export function pageDelete(page_id: string): Promise<PageDeleteResult> {
  return mutate<PageDeleteResult>(tool("bc_page_delete"), { row: { page_id } });
}

export function templateUpsert(yaml: string): Promise<MutationResult> {
  return mutate<MutationResult>(tool("bc_template_upsert"), { yaml });
}

export function labelRender(device_id: string): Promise<LabelRender> {
  return callTool<LabelRender>(tool("bc_label_render"), { device_id });
}

/* --------------------------------- reads --------------------------------- */

export function listDevices(
  params: { site_id?: string; status?: string; limit?: number } = {},
): Promise<ReadonlyArray<DeviceRow>> {
  return fetchTemplate<DeviceRow>(tool("bc_devices_list"), params);
}

export function listSites(limit = 200): Promise<ReadonlyArray<SiteRow>> {
  return fetchTemplate<SiteRow>(tool("bc_sites_list"), { limit });
}

export function listLocations(
  params: { site_id?: string; limit?: number } = {},
): Promise<ReadonlyArray<LocationRow>> {
  return fetchTemplate<LocationRow>(tool("bc_locations_list"), params);
}

export function listPages(
  params: { site_id?: string; location_id?: string; limit?: number } = {},
): Promise<ReadonlyArray<PageRow>> {
  return fetchTemplate<PageRow>(tool("bc_pages_list"), { limit: 200, ...params });
}

export function listTemplates(limit = 200): Promise<ReadonlyArray<TemplateRow>> {
  return fetchTemplate<TemplateRow>(tool("bc_templates_list"), { limit });
}

export function getTemplateYaml(template: string): Promise<ReadonlyArray<TemplateYaml>> {
  return fetchTemplate<TemplateYaml>(tool("bc_template_yaml"), { template });
}

export function pointsByDevice(device_id: string): Promise<ReadonlyArray<PointRow>> {
  return fetchTemplate<PointRow>(tool("bc_points_by_device"), { device_id });
}

export function widgetsByPage(page_id: string): Promise<ReadonlyArray<WidgetRow>> {
  return fetchTemplate<WidgetRow>(tool("bc_widgets_by_page"), { page_id });
}

export function alarmsByDevice(device_id: string): Promise<ReadonlyArray<AlarmRow>> {
  return fetchTemplate<AlarmRow>(tool("bc_alarms_by_device"), { device_id });
}

export function provisionLog(
  params: { device_id?: string; limit?: number } = {},
): Promise<ReadonlyArray<LogRow>> {
  return fetchTemplate<LogRow>(tool("bc_provision_log_recent"), params);
}

File: rubix/apps/provision-app-flutter/lib/core/api/bc_api.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import 'package:provision_app/core/network/network_providers.dart';
import 'package:provision_app/core/network/transport.dart';
import 'package:provision_app/core/api/bc_types.dart';
import 'package:provision_app/core/api/refresh.dart';

/// Typed, named wrappers over the `bc_*` tools — the Dart port of the React
/// app's `api/bc.ts`. Each call forwards to [RubixTransport.dispatch] so callers
/// never hand-spell tool ids. Reads go through `warehouse_query` (always fresh);
/// mutations bump the shared refresh signal so sibling list views re-fetch.
class BcApi {
  BcApi(this._transport, this._bumpRefresh);

  final RubixTransport _transport;
  final void Function() _bumpRefresh;

  /// The owning extension id — tool ids are `${EXTENSION_ID}.${name}`.
  static const extensionId = 'com.nubeio.rubixos';
  String _tool(String name) => '$extensionId.$name';

  /// Every write goes through here so the shared refresh signal bumps the moment
  /// the server confirms.
  Future<T> _mutate<T>(
    String toolId,
    Object? params,
    T Function(Map<String, dynamic>) parse,
  ) async {
    final res =
        await _transport.dispatch<Map<String, dynamic>>(toolId, params, fresh: true);
    _bumpRefresh();
    return parse(res);
  }

  /// A list read: `warehouse_query` envelope, always fresh.
  Future<List<R>> _query<R>(
    String template,
    R Function(Map<String, dynamic>) parse, {
    Map<String, dynamic> params = const {},
  }) async {
    final res = await _transport.dispatch<Map<String, dynamic>>(
      _tool('warehouse_query'),
      {'template': template, 'params': params},
      fresh: true,
    );
    final rows = (res['rows'] as List?) ?? const [];
    return rows
        .cast<Map<String, dynamic>>()
        .map(parse)
        .toList(growable: false);
  }

  // ── mutations ────────────────────────────────────────────────────────────

  /// read-only — no refresh bump
  Future<ScannedIdentity> decode(String barcode) async {
    final res = await _transport.dispatch<Map<String, dynamic>>(
      _tool('bc_decode'),
      {'barcode': barcode},
    );
    return ScannedIdentity.fromJson(res);
  }

  Future<ProvisionResult> provision(ProvisionInput input) =>
      _mutate(_tool('bc_provision'), input.toJson(), ProvisionResult.fromJson);

  /// Assign a (pending) device to a page — existing pageId or a newPage name.
  Future<AssignPageResult> assignPage({
    required String deviceId,
    String? pageId,
    String? newPage,
  }) {
    final params = <String, dynamic>{
      'device_id': deviceId,
      if (pageId != null) 'page_id': pageId,
      if (newPage != null) 'new_page': {'name': newPage},
    };
    return _mutate(
      _tool('bc_device_assign_page'),
      params,
      AssignPageResult.fromJson,
    );
  }

  Future<void> deviceUpdate(Map<String, dynamic> row) async {
    await _mutate(_tool('bc_device_update'), {'row': row}, (j) => j);
  }

  Future<void> decommission(List<String> deviceIds, {bool hard = false}) async {
    await _mutate(
      _tool('bc_device_decommission'),
      {'device_ids': deviceIds, 'hard': hard},
      (j) => j,
    );
  }

  Future<void> siteCreate({required String siteId, required String name}) async {
    await _mutate(
      _tool('bc_site_create'),
      {'row': {'site_id': siteId, 'name': name}},
      (j) => j,
    );
  }

  Future<void> locationCreate({
    required String locationId,
    required String siteId,
    required String name,
  }) async {
    await _mutate(
      _tool('bc_location_create'),
      {'row': {'location_id': locationId, 'site_id': siteId, 'name': name}},
      (j) => j,
    );
  }

  Future<void> pageCreate({
    required String pageId,
    required String siteId,
    required String name,
  }) async {
    await _mutate(
      _tool('bc_page_create'),
      {'row': {'page_id': pageId, 'site_id': siteId, 'name': name}},
      (j) => j,
    );
  }

  Future<void> templateUpsert(String yaml) async {
    await _mutate(_tool('bc_template_upsert'), {'yaml': yaml}, (j) => j);
  }

  Future<LabelRender> labelRender(String deviceId) async {
    final res = await _transport.dispatch<Map<String, dynamic>>(
      _tool('bc_label_render'),
      {'device_id': deviceId},
    );
    return LabelRender.fromJson(res);
  }

  // ── reads ──────────────────────────────────────────────────────────────

  Future<List<DeviceRow>> devicesList({
    String? siteId,
    String? status,
    int? limit,
  }) =>
      _query(_tool('bc_devices_list'), DeviceRow.fromJson, params: {
        if (siteId != null) 'site_id': siteId,
        if (status != null) 'status': status,
        if (limit != null) 'limit': limit,
      });

  Future<List<SiteRow>> sitesList({int limit = 200}) =>
      _query(_tool('bc_sites_list'), SiteRow.fromJson, params: {'limit': limit});

  Future<List<LocationRow>> locationsList({String? siteId, int? limit}) =>
      _query(_tool('bc_locations_list'), LocationRow.fromJson, params: {
        if (siteId != null) 'site_id': siteId,
        if (limit != null) 'limit': limit,
      });

  Future<List<PageRow>> pagesList({String? siteId, int limit = 200}) =>
      _query(_tool('bc_pages_list'), PageRow.fromJson, params: {
        if (siteId != null) 'site_id': siteId,
        'limit': limit,
      });

  Future<List<TemplateRow>> templatesList({int limit = 200}) => _query(
        _tool('bc_templates_list'),
        TemplateRow.fromJson,
        params: {'limit': limit},
      );

  Future<List<TemplateYaml>> templateYaml(String template) => _query(
        _tool('bc_template_yaml'),
        TemplateYaml.fromJson,
        params: {'template': template},
      );

  Future<List<PointRow>> pointsByDevice(String deviceId) => _query(
        _tool('bc_points_by_device'),
        PointRow.fromJson,
        params: {'device_id': deviceId},
      );

  Future<List<WidgetRow>> widgetsByPage(String pageId) => _query(
        _tool('bc_widgets_by_page'),
        WidgetRow.fromJson,
        params: {'page_id': pageId},
      );
}

/// The app-wide [BcApi], wired to the transport and the refresh signal.
final bcApiProvider = Provider<BcApi>((ref) {
  final transport = ref.watch(transportProvider);
  return BcApi(transport, () => ref.read(refreshProvider.notifier).bump());
});
